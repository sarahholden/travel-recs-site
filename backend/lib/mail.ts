import 'dotenv/config';
import {
  createTransport,
  getTestMessageUrl,
  SentMessageInfo,
} from 'nodemailer';

const transport = createTransport({
  host: process.env.MAIL_HOST,
  port: 587,
  auth: {
    user: process.env.MAIL_USER,
    pass: process.env.MAIL_PASS,
  },
});

function makeANiceEmail(text: string) {
  return `
    <div style="
      border: 1px solid black;
      padding: 20px;
      font-family: sans-serif;
      line-height: 2;
      font-size: 20px;
    ">
      <h2>Hello There!</h2>
      <p>${text}</p>
      <p>¬©Ô∏è, Sarah Holden</p>
    </div>
  `;
}

type MailResponse = {
  accepted?: string[] | null;
  rejected?: string[] | null;
  envelopeTime: number;
  messageTime: number;
  messageSize: number;
  response: string;
  envelope: {
    from: string;
    to?: string[] | null;
  };
  messageId: string;
};

export async function sendPasswordResetEmail(
  resetToken: string,
  to: string
): Promise<void> {
  const info = (await transport.sendMail({
    to,
    from: 'test@example.com',
    subject: 'Your Sick Fits password reset token!',
    html: makeANiceEmail(`Your Password reset token is here!
      <a href="${process.env.FRONTEND_URL}/reset?token=${resetToken}">Click here to reset</a>
    `),
  })) as MailResponse;

  if (process.env.MAIL_USER.includes('ethereal.email')) {
    /* eslint no-console: 0 */
    console.log(
      'üì¨ Message Sent! Preview it at',
      getTestMessageUrl(info as SentMessageInfo)
    );
  }
}
