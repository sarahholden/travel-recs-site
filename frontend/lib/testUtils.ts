import casual from "casual";
// import { PAGINATION_QUERY } from '../components/Pagination';

// seed it so we get consistent results
casual.seed(777);

const fakeGuide = () => ({
  // __typename: 'Item',
  id: "abc123",
  user: null,
  altText: "New York City Skyline",
  destinations: [],
  image: {
    publicUrlTransformed: "dog.jpg",
  },
  location_info: "The biggest apple in existance",
  name: "New York City",
  status: "AVAILABLE",
});

const fakeUser = (overrides) => ({
  __typename: "User",
  id: "4234",
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: "Order",
  id: "ord123",
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2022-12-11T20:16:13.797Z",
  user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
  __typename: "CartItem",
  id: "omg123",
  quantity: 3,
  product: fakeGuide(),
  user: fakeUser(),
  ...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: "aggregate",
            aggregate: {
              count: length,
              __typename: "count",
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  LocalStorageMock,
  fakeGuide,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
