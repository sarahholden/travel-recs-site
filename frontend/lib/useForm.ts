import { useState } from "react"

type GuideInputsType = {
  name?: string;
  location_info?: string;
}

type ElementPropsType = {
  name: string;
  type: string;
  value: string | number | File
}

export default function useForm(initial: GuideInputsType = {}) {
  const [inputs, setInputs] = useState(initial)

  
  function handleChange(e: React.ChangeEvent<HTMLInputElement> & React.ChangeEvent<HTMLTextAreaElement>) {
    let {name, type, value}: ElementPropsType = e.target;

    if (type === 'number') {
      value = parseInt(value)
    }

    if (type === 'file' && e?.target.files) {
      value = e.target.files[0]
    }

    setInputs({
      ...inputs,
      [name]: value
    })
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.fromEntries(Object.entries(inputs).map(([key, value]) => [key, '']))
    setInputs(blankState)
  }

  return { inputs, handleChange, resetForm, clearForm }
  
}