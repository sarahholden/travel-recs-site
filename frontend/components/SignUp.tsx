import Router from "next/router";
import useForm from "../lib/useForm";
import { refetchUserQuery, useSignUpMutation } from "../types/generated-queries";
import { Form } from "./styles/FormStyles";
import DisplayError from "./ErrorMessage";
import Link from "next/link";
import SignIn from "./SignIn";

export default function SignUp() {
  const {inputs, handleChange, resetForm} = useForm({
    email: '',
    password: '',
    name: '',
  });

  const [signUpMutation, {data, loading, error}] = useSignUpMutation()

  async function handleSubmit(e: React.FormEvent) {
    e.preventDefault();

    // TODO - add errors to form fields
    if (!inputs.email || !inputs.password || !inputs.name) return;

    const res = await signUpMutation({
      variables: {
        email: inputs.email,
        password: inputs.password,
        name: inputs.name
      },
      refetchQueries: [refetchUserQuery()],
    })
    resetForm()
  }

  return (
    <>
    {error && <DisplayError error={error} />}


      {data?.createUser ? (
        <div>
          <h2>Sign In to your account</h2>
          Signed up with {data.createUser.email}. Sign in to go to your account.
          <SignIn/>
        </div>
      ) : (
        <>
          <h2>Sign Up for an account</h2>
          <Form method="POST" onSubmit={handleSubmit}>

            <fieldset disabled={loading} aria-busy={loading}>
              <div>
                <label htmlFor="email">
                  Email Address
                </label>
                <input 
                  type="email"
                  id="email"
                  name="email"
                  placeholder="Enter Your Email"
                  value={inputs.email}
                  autoComplete="email"
                  required
                  onChange={handleChange} />
              </div>
              <div>
                <label htmlFor="name">
                  Name
                </label>
                <input 
                  type="text"
                  id="name"
                  name="name"
                  placeholder="Enter Your Name"
                  autoComplete="name"
                  value={inputs.name}
                  required
                  onChange={handleChange} />
              </div>
              <div>
                <label htmlFor="password">
                  Password
                </label>
                <input 
                  type="password"
                  id="password"
                  name="password"
                  placeholder="Enter Your password"
                  autoComplete="password"
                  value={inputs.password}
                  required
                  onChange={handleChange} />
              </div>
              <div>
                <button type="submit">Sign Up</button>
              </div>
              <div>
                Already have an account? <Link href="signin">Sign In</Link>
              </div>
            </fieldset>
          </Form>
        </>
      )
    }

    </>
  )
}