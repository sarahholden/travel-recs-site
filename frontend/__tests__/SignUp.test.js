import { findByText, render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import userEvent from "@testing-library/user-event";
import { SignUpDocument, UserDocument } from "../types/generated-queries";
import SignUp from "../components/SignUp";
import { fakeUser } from "../lib/testUtils";

const me = fakeUser();
const mePassword = 'sarah';

const mocks = [
  {
    request: {
      query: SignUpDocument,
      variables: {
        name: me.name,
        email: me.email,
        password: mePassword,
      },
    },
    result: {
      data: {
        createUser: {
          __typename: "User",
          id: "abc123",
          email: me.email,
          name: me.name,
        },
      },
    },
  },
  {
    request: { query: UserDocument },
    result: { data: { authenticatedItem: me } },
  },
];


describe('<SignUp/>', () => {
  it("Renders and matches snapshot", async () => {
    const { container } = render(
      <MockedProvider>
        <SignUp />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();

  });
  it("Renders and matches snapshot", async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <SignUp />
      </MockedProvider>
    );

 ;
    await userEvent.type(screen.getByLabelText(/name/i), me.name);
    await userEvent.type(screen.getByLabelText(/email address/i), me.email);
    await userEvent.type(screen.getByLabelText(/password/i), mePassword);
    await userEvent.click(
      screen.getByRole("button", {
        name: /sign up/i,
      })
    );
    await screen.findByText(/Signed up with/i);
  });
})