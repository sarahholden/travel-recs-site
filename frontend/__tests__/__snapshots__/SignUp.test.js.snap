// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUp/> Renders and matches snapshot 1`] = `
<div>
  <h2>
    Sign Up for an account
  </h2>
  <form
    class="FormStyles__Form-sc-sf5fob-0 KLHJ"
    method="POST"
  >
    <fieldset
      aria-busy="false"
    >
      <div>
        <label
          for="email"
        >
          Email Address
        </label>
        <input
          autocomplete="email"
          id="email"
          name="email"
          placeholder="Enter Your Email"
          required=""
          type="email"
          value=""
        />
      </div>
      <div>
        <label
          for="name"
        >
          Name
        </label>
        <input
          autocomplete="name"
          id="name"
          name="name"
          placeholder="Enter Your Name"
          required=""
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="password"
          id="password"
          name="password"
          placeholder="Enter Your password"
          required=""
          type="password"
          value=""
        />
      </div>
      <div>
        <button
          type="submit"
        >
          Sign Up
        </button>
      </div>
      <div>
        Already have an account? 
        <a
          href="/signin"
        >
          Sign In
        </a>
      </div>
    </fieldset>
  </form>
</div>
`;
